File: binding/include/libqslider.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qslider.html
pub const qslider = struct {

    /// New constructs a new QSlider object.
    pub fn New(parent: ?*anyopaque) QtC.QSlider {
        return qtc.QSlider_new(@ptrCast(parent));
    }


    /// New2 constructs a new QSlider object.
    pub fn New2() QtC.QSlider {
        return qtc.QSlider_new2();
    }


    /// New3 constructs a new QSlider object.
    pub fn New3(orientation: i32) QtC.QSlider {
        return qtc.QSlider_new3(@intCast(orientation));
    }


    /// New4 constructs a new QSlider object.
    pub fn New4(orientation: i32, parent: ?*anyopaque) QtC.QSlider {
        return qtc.QSlider_new4(@intCast(orientation), @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSlider_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qslider.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QSlider_SizeHint(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnSizeHint(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QSize) void {
        qtc.QSlider_OnSizeHint(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QSlider_QBaseSizeHint(@ptrCast(self));
    }

    pub fn MinimumSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QSlider_MinimumSizeHint(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnMinimumSizeHint(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QSize) void {
        qtc.QSlider_OnMinimumSizeHint(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMinimumSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QSlider_QBaseMinimumSizeHint(@ptrCast(self));
    }

    pub fn SetTickPosition(self: ?*anyopaque, position: i32) void {
        qtc.QSlider_SetTickPosition(@ptrCast(self), @intCast(position));
    }

    pub fn TickPosition(self: ?*anyopaque, ) i32 {
        return qtc.QSlider_TickPosition(@ptrCast(self));
    }

    pub fn SetTickInterval(self: ?*anyopaque, ti: i32) void {
        qtc.QSlider_SetTickInterval(@ptrCast(self), ti);
    }

    pub fn TickInterval(self: ?*anyopaque, ) i32 {
        return qtc.QSlider_TickInterval(@ptrCast(self));
    }

    pub fn Event(self: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QSlider_Event(@ptrCast(self), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QSlider_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QSlider_QBaseEvent(@ptrCast(self), @ptrCast(event));
    }

    pub fn PaintEvent(self: ?*anyopaque, ev: ?*anyopaque) void {
        qtc.QSlider_PaintEvent(@ptrCast(self), @ptrCast(ev));
    }

    /// Allows for overriding the related default method
    pub fn OnPaintEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QSlider_OnPaintEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePaintEvent(self: ?*anyopaque, ev: ?*anyopaque) void {
        qtc.QSlider_QBasePaintEvent(@ptrCast(self), @ptrCast(ev));
    }

    pub fn MousePressEvent(self: ?*anyopaque, ev: ?*anyopaque) void {
        qtc.QSlider_MousePressEvent(@ptrCast(self), @ptrCast(ev));
    }

    /// Allows for overriding the related default method
    pub fn OnMousePressEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QSlider_OnMousePressEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMousePressEvent(self: ?*anyopaque, ev: ?*anyopaque) void {
        qtc.QSlider_QBaseMousePressEvent(@ptrCast(self), @ptrCast(ev));
    }

    pub fn MouseReleaseEvent(self: ?*anyopaque, ev: ?*anyopaque) void {
        qtc.QSlider_MouseReleaseEvent(@ptrCast(self), @ptrCast(ev));
    }

    /// Allows for overriding the related default method
    pub fn OnMouseReleaseEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QSlider_OnMouseReleaseEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMouseReleaseEvent(self: ?*anyopaque, ev: ?*anyopaque) void {
        qtc.QSlider_QBaseMouseReleaseEvent(@ptrCast(self), @ptrCast(ev));
    }

    pub fn MouseMoveEvent(self: ?*anyopaque, ev: ?*anyopaque) void {
        qtc.QSlider_MouseMoveEvent(@ptrCast(self), @ptrCast(ev));
    }

    /// Allows for overriding the related default method
    pub fn OnMouseMoveEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QSlider_OnMouseMoveEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMouseMoveEvent(self: ?*anyopaque, ev: ?*anyopaque) void {
        qtc.QSlider_QBaseMouseMoveEvent(@ptrCast(self), @ptrCast(ev));
    }

    pub fn InitStyleOption(self: ?*anyopaque, option: ?*anyopaque) void {
        qtc.QSlider_InitStyleOption(@ptrCast(self), @ptrCast(option));
    }

    /// Allows for overriding the related default method
    pub fn OnInitStyleOption(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QSlider_OnInitStyleOption(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseInitStyleOption(self: ?*anyopaque, option: ?*anyopaque) void {
        qtc.QSlider_QBaseInitStyleOption(@ptrCast(self), @ptrCast(option));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSlider_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qslider.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSlider_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qslider.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSlider_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/slider.html#types
pub const enums = struct {
    pub const TickPosition = enum {
        pub const NoTicks: i32 = 0;
        pub const TicksAbove: i32 = 1;
        pub const TicksLeft: i32 = 1;
        pub const TicksBelow: i32 = 2;
        pub const TicksRight: i32 = 2;
        pub const TicksBothSides: i32 = 3;
    };

};

File: binding/src/libqslider.cpp
#include <QAbstractSlider>
#include <QEvent>
#include <QMouseEvent>
#include <QPaintEvent>
#include <QSize>
#include <QSlider>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QStyleOptionSlider>
#include <QWidget>
#include <qslider.h>
#include "libqslider.h"
#include "libqslider.hxx"

QSlider* QSlider_new(QWidget* parent) {
	 return new VirtualQSlider(parent);
}

QSlider* QSlider_new2() {
	 return new VirtualQSlider();
}

QSlider* QSlider_new3(int orientation) {
	 return new VirtualQSlider(static_cast<Qt::Orientation>(orientation));
}

QSlider* QSlider_new4(int orientation, QWidget* parent) {
	 return new VirtualQSlider(static_cast<Qt::Orientation>(orientation), parent);
}

libqt_string QSlider_Tr(const char* s) {
	QString _ret = QSlider::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QSize* QSlider_SizeHint(const QSlider* self) {
	return new QSize(self->sizeHint());
}

QSize* QSlider_MinimumSizeHint(const QSlider* self) {
	return new QSize(self->minimumSizeHint());
}

void QSlider_SetTickPosition(QSlider* self, int position) {
	self->setTickPosition(static_cast<QSlider::TickPosition>(position));
}

int QSlider_TickPosition(const QSlider* self) {
	return self->tickPosition();
}

void QSlider_SetTickInterval(QSlider* self, int ti) {
	self->setTickInterval(ti);
}

int QSlider_TickInterval(const QSlider* self) {
	return self->tickInterval();
}

bool QSlider_Event(QSlider* self, QEvent* event) {
	return self->event(event);
}

libqt_string QSlider_Tr2(const char* s, const char* c) {
	QString _ret = QSlider::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSlider_Tr3(const char* s, const char* c, int n) {
	QString _ret = QSlider::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
QSize* QSlider_QBaseSizeHint(const QSlider* self) {
	auto* vqslider = dynamic_cast<const VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_SizeHint_IsBase(true);
	return new QSize(vqslider->sizeHint());
}
}

// Auxiliary method to allow providing re-implementation
void QSlider_OnSizeHint(const QSlider* self, intptr_t slot) {
	auto* vqslider = dynamic_cast<const VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_SizeHint_Callback(reinterpret_cast<VirtualQSlider::QSlider_SizeHint_Callback>(slot));
}
}

// Base class handler implementation
QSize* QSlider_QBaseMinimumSizeHint(const QSlider* self) {
	auto* vqslider = dynamic_cast<const VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_MinimumSizeHint_IsBase(true);
	return new QSize(vqslider->minimumSizeHint());
}
}

// Auxiliary method to allow providing re-implementation
void QSlider_OnMinimumSizeHint(const QSlider* self, intptr_t slot) {
	auto* vqslider = dynamic_cast<const VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_MinimumSizeHint_Callback(reinterpret_cast<VirtualQSlider::QSlider_MinimumSizeHint_Callback>(slot));
}
}

// Base class handler implementation
bool QSlider_QBaseEvent(QSlider* self, QEvent* event) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_Event_IsBase(true);
	return vqslider->event(event);
}
}

// Auxiliary method to allow providing re-implementation
void QSlider_OnEvent(QSlider* self, intptr_t slot) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_Event_Callback(reinterpret_cast<VirtualQSlider::QSlider_Event_Callback>(slot));
}
}

// Derived class handler implementation
void QSlider_PaintEvent(QSlider* self, QPaintEvent* ev) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
	vqslider->paintEvent(ev);
	} else {
	self->QSlider::paintEvent(ev);
}
}

// Base class handler implementation
void QSlider_QBasePaintEvent(QSlider* self, QPaintEvent* ev) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_PaintEvent_IsBase(true);
	vqslider->paintEvent(ev);
}
}

// Auxiliary method to allow providing re-implementation
void QSlider_OnPaintEvent(QSlider* self, intptr_t slot) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_PaintEvent_Callback(reinterpret_cast<VirtualQSlider::QSlider_PaintEvent_Callback>(slot));
}
}

// Derived class handler implementation
void QSlider_MousePressEvent(QSlider* self, QMouseEvent* ev) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
	vqslider->mousePressEvent(ev);
	} else {
	self->QSlider::mousePressEvent(ev);
}
}

// Base class handler implementation
void QSlider_QBaseMousePressEvent(QSlider* self, QMouseEvent* ev) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_MousePressEvent_IsBase(true);
	vqslider->mousePressEvent(ev);
}
}

// Auxiliary method to allow providing re-implementation
void QSlider_OnMousePressEvent(QSlider* self, intptr_t slot) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_MousePressEvent_Callback(reinterpret_cast<VirtualQSlider::QSlider_MousePressEvent_Callback>(slot));
}
}

// Derived class handler implementation
void QSlider_MouseReleaseEvent(QSlider* self, QMouseEvent* ev) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
	vqslider->mouseReleaseEvent(ev);
	} else {
	self->QSlider::mouseReleaseEvent(ev);
}
}

// Base class handler implementation
void QSlider_QBaseMouseReleaseEvent(QSlider* self, QMouseEvent* ev) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_MouseReleaseEvent_IsBase(true);
	vqslider->mouseReleaseEvent(ev);
}
}

// Auxiliary method to allow providing re-implementation
void QSlider_OnMouseReleaseEvent(QSlider* self, intptr_t slot) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_MouseReleaseEvent_Callback(reinterpret_cast<VirtualQSlider::QSlider_MouseReleaseEvent_Callback>(slot));
}
}

// Derived class handler implementation
void QSlider_MouseMoveEvent(QSlider* self, QMouseEvent* ev) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
	vqslider->mouseMoveEvent(ev);
	} else {
	self->QSlider::mouseMoveEvent(ev);
}
}

// Base class handler implementation
void QSlider_QBaseMouseMoveEvent(QSlider* self, QMouseEvent* ev) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_MouseMoveEvent_IsBase(true);
	vqslider->mouseMoveEvent(ev);
}
}

// Auxiliary method to allow providing re-implementation
void QSlider_OnMouseMoveEvent(QSlider* self, intptr_t slot) {
	auto* vqslider = dynamic_cast<VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_MouseMoveEvent_Callback(reinterpret_cast<VirtualQSlider::QSlider_MouseMoveEvent_Callback>(slot));
}
}

// Derived class handler implementation
void QSlider_InitStyleOption(const QSlider* self, QStyleOptionSlider* option) {
	auto* vqslider = dynamic_cast<const VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
	vqslider->initStyleOption(option);
	} else {
	self->QSlider::initStyleOption(option);
}
}

// Base class handler implementation
void QSlider_QBaseInitStyleOption(const QSlider* self, QStyleOptionSlider* option) {
	auto* vqslider = dynamic_cast<const VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_InitStyleOption_IsBase(true);
	vqslider->initStyleOption(option);
}
}

// Auxiliary method to allow providing re-implementation
void QSlider_OnInitStyleOption(const QSlider* self, intptr_t slot) {
	auto* vqslider = dynamic_cast<const VirtualQSlider*>(self);
	if (vqslider && vqslider->isVirtualQSlider) {
vqslider->setQSlider_InitStyleOption_Callback(reinterpret_cast<VirtualQSlider::QSlider_InitStyleOption_Callback>(slot));
}
}

void QSlider_Delete(QSlider* self) {
    delete self;
}

